webapp/src/ts/modules/reports/reports.component.ts

```diff
@@ -178,23 +177,34 @@ export class ReportsComponent implements OnInit, AfterViewInit, OnDestroy {
     if (report.validSubject) {
       return report.subject.value;
     }
-    if (report.subject.name) {
+    if (report.subject?.name) {
       return report.subject.name;
     }
     return this.translateService.instant('report.subject.unknown');
   }
 
   private prepareReports(reports) {
-    return reports.map(report => {
-      const form = _find(this.forms, { code: report.form });
-      report.icon = form && form.icon;
-      report.heading = this.getReportHeading(form, report);
-      report.summary = form ? form.title : report.form;
-      report.lineage = report.subject && report.subject.lineage || report.lineage;
-      report.unread = !report.read;
-
-      return report;
-    });
+    return this.currentLevel.then(
+      currentLevel => {
+        return reports.map(report => {
+          const form = _find(this.forms, { code: report.form });
+          report.icon = form && form.icon;
+          report.heading = this.getReportHeading(form, report);
+          report.summary = form ? form.title : report.form;
+          report.lineage = report.subject && report.subject.lineage || report.lineage;
+          // remove the lineage level that belongs to the offline logged-in user
+          if (currentLevel && report.lineage && report.lineage.length) {
+            report.lineage = report.lineage.filter(level => level);
+            if(report.lineage[report.lineage.length-1] === currentLevel){
+              report.lineage.pop();
+            }
+          }
+          report.unread = !report.read;
+          return report;
+        });
+      }
+    );
```

Comment:
it's always always better to keep code flat and check for null, this version adds validation when getting the last element of the array

```suggestion
  const userLineageLevel = await this.currentLevel;
  
  return reports.map(report => {
  const form = _find(this.forms, { code: report.form });
  report.icon = form && form.icon;
  report.heading = this.getReportHeading(form, report);
  report.summary = form ? form.title : report.form;
  report.lineage = report.subject && report.subject.lineage || report.lineage;
  report.unread = !report.read;
          
  // remove the lineage level that belongs to the offline logged-in user
  if (currentLevel && report?.lineage?.length) {
    report.lineage = report.lineage.filter(level => level);
    const item = report.lineage?.length && report.lineage[report.lineage.length -1]
    if (item === userLineageLevel) {
      report.lineage.pop();
    }
  }
  
  return report;
 );
```

Reply:
well noted.